runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkBeamMetric.java

```diff
@@ -56,7 +56,10 @@
 
   private String renderName(MetricResult<?> metricResult) {
     MetricName metricName = metricResult.name();
-    String rendered = metricResult.step() + "." + metricName.namespace() + "." + metricName.name();
+    String rendered =
+        metricResult.step().replace(".", "_")
```

Comment:
Might be worth checking if `metricResult.step().replace(".", "_").endsWith("_")`, and if so remove it as well.